src/core/deltafilewrapper.cpp

```diff
@@ -487,14 +487,16 @@ QMap<QString, QString> DeltaFileWrapper::attachmentFileNames() const
 
 void DeltaFileWrapper::addPatch( const QString &localLayerId, const QString &sourceLayerId, const QString &localPkAttrName, const QString &sourcePkAttrName, const QgsFeature &oldFeature, const QgsFeature &newFeature, bool storeSnapshot )
 {
+  QVariant localPk = oldFeature.attribute( localPkAttrName );
+  QVariant sourcePk = oldFeature.attribute( sourcePkAttrName );
   QJsonObject delta(
     {
-      { "localPk", oldFeature.attribute( localPkAttrName ).toString() },
+      { "localPk", localPk.isNull() ? QString() : localPk.toString() },
```

Comment:
Safer to use QgsVariantUtils::isNull to do a null check against a QVariant (for all t he isNull() added in this PR)